src/components/WalletSelector/AddWalletPending/AddWalletPendingGnosisSafe.tsx

```diff
@@ -54,69 +62,140 @@ function AddWalletPendingGnosisSafe({
-      setDetectedError,
-      library,
-    ]
+    [pendingWallet, library, authenticateWithBackend, handleError]
   );
 
+  // Validates the signature on-chain. If it hasnt been signed yet, initializes a listener to wait for the SignMsg event.
+  // This is used in 2 cases:
+  // 1. The user has previously tried to sign a message, and they opted to retry using the same nonce+transaction
+  // 2. This gets automatically called on an interval as a backup to validate the signature in case the listener fails to detect the SignMsg event
+  const manuallyValidateSignature = useCallback(async () => {
+    if (!account) {
+      return;
+    }
+
+    try {
+      // Immediately check if the message has already been signed and executed on chain
+      const wasSigned = await validateNonceSignedByGnosis(account, nonce, library);
+      if (wasSigned) {
+        await authenticateWithBackend(account, nonce);
+      }
+
+      // If it hasn't, set up a listener because the transaction may not have been executed yet
+      if (pendingState !== LISTENING_ONCHAIN) {
+        setPendingState(LISTENING_ONCHAIN);
+        await listenForGnosisSignature(account, nonce, library);
+        // Once signed, call the backend as usual
+        void authenticateWithBackend(account, nonce);
+      }
+    } catch (error: unknown) {
+      handleError(error);
+    }
+  }, [account, authenticateWithBackend, handleError, library, pendingState, nonce]);
+
+  const restartAuthentication = useCallback(async () => {
+    if (account) {
+      await attemptAddWallet(account.toLowerCase(), nonce, userExists);
+    }
+  }, [account, attemptAddWallet, nonce, userExists]);
+
+  // This runs once to auto-initiate the authentication flow, when wallet is first connected (ie when 'account' is defined)
   useEffect(() => {
-    async function authenticate() {
+    if (authenticationFlowStarted) {
+      return;
+    }
+
+    async function initiateAuthentication() {
       if (account) {
-        if (authenticatedUserAddresses.includes(account.toLowerCase())) {
-          setPendingState(ADDRESS_ALREADY_CONNECTED);
-          return;
-        }
+        setAuthenticationFlowStarted(true);
+
+        try {
+          if (authenticatedUserAddresses.includes(account.toLowerCase())) {
+            setPendingState(ADDRESS_ALREADY_CONNECTED);
+            return;
+          }
+
+          const { nonce, user_exists: userExists } = await fetchNonce(account, fetcher);
+          setNonce(nonce);
+          setUserExists(userExists);
 
-        await attemptAddWallet(account.toLowerCase());
+          if (nonce === previousAttemptNonce) {
+            return;
+          }
+
+          await attemptAddWallet(account.toLowerCase(), nonce, userExists);
+        } catch (error: unknown) {
+          handleError(error);
+        }
       }
     }
 
-    void authenticate();
-  }, [account, authenticatedUserAddresses, attemptAddWallet]);
+    void initiateAuthentication();
+  }, [
```

Comment:
may not be necessary, but are there any threads/processes that are being opened as a result of this `useEffect`? if so it's best practice to clean them up as a return func, but not sure if needed here

```js
useEffect(() => {
  // do stuff

  return () => {} // clean stuff up
}, [])
```

Reply:
will fast follow with optimizations here as discussed